### PR TITLE
Add unit tests for deployment pkg Pull func

### PR DESCRIPTION
**Pull request by sebrandon1**
Builds on top of #261 

Proof of concept PR for adding unit tests wrapping some of the deployment package starting with the `Pull` func.

Notable changes:
- Added func `GetTestClients` which helps build a test client for doing unit testing.
- Changed `settings.K8sClient` to use the Interface instead of the client.
- Changed `settings.NetworkingV1Client` to `settings.NetworkingV1Inteface`.
- Vendor changes come from using the fake k8s client and the additional code needed for that as well as the stretchr/testify library.
- Changes `v1` to `appsv1` similar to #239

### CODE REVIEW

**Comment by sebrandon1** [comment]
Once #273 merges I'll rebase this PR on top of the changes there.

File: pkg/deployment/deployment.go [hunk 1]
```diff
@@ -101,14 +101,20 @@ func Pull(apiClient *clients.Settings, name, nsname string) (*Builder, error) {
...
 		builder.errorMsg = "deployment 'name' cannot be empty"
+
+		return &builder, fmt.Errorf(builder.errorMsg)
 	}
 
 	if nsname == "" {
 		builder.errorMsg = "deployment 'namespace' cannot be empty"
+
+		return &builder, fmt.Errorf(builder.errorMsg)
 	}
```
Comment (kononovn) [inline pkg/deployment/deployment.go:113]: Please add single verification and remove multiple return above.

	if builder.errorMsg != "" {
		return nil, fmt.Errorf("faield to pull deployment object due to the following error: %s", builder.errorMsg)
	}

Comment (sebrandon1) [reply pkg/deployment/deployment.go:113]: Ah yes I forgot to change this after we talked about the builder errorMsg structure.  I'll change that.

File: pkg/deployment/deployment.go [hunk 2]
```diff
@@ -101,14 +101,20 @@ func Pull(apiClient *clients.Settings, name, nsname string) (*Builder, error) {
...
 
 	if !builder.Exists() {
-		return nil, fmt.Errorf("deployment object %s doesn't exist in namespace %s", name, nsname)
+		builder.errorMsg = fmt.Sprintf("deployment object %s doesn't exist in namespace %s", name, nsname)
```
Comment (kononovn) [inline pkg/deployment/deployment.go:115]: I don't think we need to change it. We failed to pull deployment let's return nil instead of invalid builder.

**Review by kononovn — CHANGES_REQUESTED** [review]
Please check few comments above.

File: pkg/deployment/deployment.go (hunk 2 above)
Comment (sebrandon1) [reply pkg/deployment/deployment.go]: Done.

File: pkg/deployment/deployment.go (hunk 1 above)
Comment (sebrandon1) [reply pkg/deployment/deployment.go:113]: I think the pattern should be if you set an `errorMsg`, you should immediately return.  Let's discuss!

Comment (kononovn) [reply pkg/deployment/deployment.go:113]: ok we can use multiple returns

File: pkg/deployment/deployment.go [hunk 3]
```diff
@@ -101,10 +101,14 @@ func Pull(apiClient *clients.Settings, name, nsname string) (*Builder, error) {
 
 	if name == "" {
 		builder.errorMsg = "deployment 'name' cannot be empty"
+
+		return &builder, fmt.Errorf(builder.errorMsg)
```
Comment (kononovn) [inline pkg/deployment/deployment.go:105]: Please return nil instead of builder

File: pkg/deployment/deployment.go [hunk 4]
```diff
@@ -101,10 +101,14 @@ func Pull(apiClient *clients.Settings, name, nsname string) (*Builder, error) {
...
 	}
 
 	if nsname == "" {
 		builder.errorMsg = "deployment 'namespace' cannot be empty"
+
+		return &builder, fmt.Errorf(builder.errorMsg)
```
Comment (kononovn) [inline pkg/deployment/deployment.go:111]: return &builder, fmt.Errorf("deployment 'namespace' cannot be empty")

**Review by kononovn — CHANGES_REQUESTED** [review]
2 more comments

File: pkg/deployment/deployment.go [hunk 5]
```diff
@@ -101,10 +101,14 @@ func Pull(apiClient *clients.Settings, name, nsname string) (*Builder, error) {
 
 	if name == "" {
 		builder.errorMsg = "deployment 'name' cannot be empty"
+
+		return nil, fmt.Errorf(builder.errorMsg)
 	}
 
 	if nsname == "" {
 		builder.errorMsg = "deployment 'namespace' cannot be empty"
```
Comment (kononovn) [inline pkg/deployment/deployment.go:109]: This line can be safely removed

File: pkg/deployment/deployment.go [hunk 6]
```diff
@@ -101,10 +101,14 @@ func Pull(apiClient *clients.Settings, name, nsname string) (*Builder, error) {
 
 	if name == "" {
 		builder.errorMsg = "deployment 'name' cannot be empty"
```
Comment (kononovn) [inline pkg/deployment/deployment.go:103]: This line can be safely removed

**Review by kononovn — APPROVED** [review]
/lgtm

File: pkg/deployment/deployment.go (hunk 5 above)
Comment (sebrandon1) [reply pkg/deployment/deployment.go]: Done.

File: pkg/deployment/deployment.go (hunk 3 above)
Comment (sebrandon1) [reply pkg/deployment/deployment.go]: Done.

File: pkg/deployment/deployment_test.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	"github.com/stretchr/testify/assert"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+//nolint:funlen
```
Comment (achuzhoy) [inline pkg/deployment/deployment_test.go:14]: Why the exception?

File: pkg/deployment/deployment.go (hunk 4 above)
Comment (sebrandon1) [reply pkg/deployment/deployment.go]: Done.

File: pkg/deployment/deployment_test.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+		if testCase.expectedError {
+			assert.NotNil(t, err)
+
+			// Check the error message
+			if testCase.expectedErrorText != "" {
+				assert.Equal(t, testCase.expectedErrorText, err.Error())
+			}
+		} else {
+			assert.Nil(t, err)
+			assert.Equal(t, testDeployment.Name, builderResult.Object.Name)
```
Comment (trewest) [inline pkg/deployment/deployment_test.go:111]: Guess we might as well add an assertion for namespace as well

File: pkg/deployment/deployment_test.go (hunk 7 above)
Comment (sebrandon1) [reply pkg/deployment/deployment_test.go:14]: Length of the test function was too long for the linter.

File: pkg/deployment/deployment_test.go (hunk 8 above)
Comment (sebrandon1) [reply pkg/deployment/deployment_test.go:111]: Good call I can add that.

Comment (sebrandon1) [reply pkg/deployment/deployment_test.go:111]: Done.

**Review by trewest — APPROVED** [review]
/lgtm

**Pull request merged**